keps/sig-api-machinery/2211-strip-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,413 @@
+While the managedFields is beneficial to resolving conflicts on updates, it is a bit verbose to many users and controllers
+because this field could be quite large and generally users and controllers are not interested in such information. 
+In addition, the bandwidth usage and size of internal caches could be significantly affected due to the additional managedFields.
+
+This KEP is proposing stripping the managedFields on the server-side when clients request objects to make the response less verbose
+and save the amount of data that needs to be transferred.
+
+## Motivation
+
+Many users have complained in [kubernetes#90066][90066] that they got impacted by the addition of managedFields
+(and the issue itself has received more than 100 thumb-up reactions), and we want to improve the user experience.
+
+[90066]: https://github.com/kubernetes/kubernetes/issues/90066
+
+At the same time, it is not ideal to only drop the managedFields on the client-side, because that requires the same
+bandwidth usage compared to simply returning the original object, and every client
+(such as kubectl or CLI tools written by users using client-go) needs to strip the managedFields by themselves.
+
+With this KEP, the API server could return objects without managedFields on demand, so that least client-side changes is needed
+to make the response less verbose. 
+
+### Goals
+
+* Omit managedFields from responses, when requested by clients.
+
+### Non-Goals
+
+* Turn off managedFields. Server-side Apply relies on managedFields, we could not simply disable it.
+* Clear manageFields only on the client-side.
+
+## Proposal
+
+### Server-side changes
+
+* Add three media types, so that client could set the `Accept` header to them to ask the API server to strip the managedFields.
+  * `application/json+managed.omit`
+  * `application/yaml+managed.omit`
+  * `application/vnd.kubernetes.protobuf+managed.omit`
+* Add a new serializer, say `UnmanagedSerializer`, which wraps existing serializers and strips the managedFields before encoding objects.
+* Make the API server to use `UnmanagedSerializer` when clients set the `Accept` header to the above three media types.
+
+It is worth noting that we could add similar media types to represent specific format version of managedFields,
+such as `application/json+managed.v1`, `application/yaml+managed.v2` etc, so that users are able to request a  certain
+version of managedFields with these headers.
+
+### Client-side changes
+
+FIXME: we need to set the `Accept` header on the client-side and perform client-side stripping if the API server still
+returns objects with managedFields. I am not sure we should do this in `client-go` and `cli-runtime` or only in `kubectl get`.
+
+/cc @apelisse
+
+### Risks and Mitigations
+
+The API server might not be up-to-date and cannot recognize the new media types, hence fails to omit managedFields from responses.
```

Comment:
The current implementation would cause the API Server to reject the request with a 406 (not acceptable).
This is where `Accept: application/yaml+unmanaged,application/yaml;q=0.9` which fallsback on default yaml if the server is too old.
We could mention that here, but we should clarify that an older server won't send the fields but fail the request (at least it should).

Reply:
This is a good point, I'll mention that. 

Additionally I have tried locally with `KinD` and it seems that `;q=0.9` is not necessary, as long as we provide a fallback media type at the end, the API server would not reject the request.

Reply:
Yes, for fallback it doesn't need `q`, but the value of `q` is being used to sort the headers so if omitted it might cause a different header to win the selection.

Reply:
I can't find the exact spec, but I'm pretty sure they should be (and k8s does) treated in order.

Reply:
The headers get sorted before being used by the negotiator according to those rules: https://github.com/munnerz/goautoneg/blob/master/autoneg.go#L62
So it seems that the actual order they are provided in does not matter. Or am I missing something?

Reply:
Yeah, it looks like it's moving the highest Qs first in line, which seems fine. On the other hand, it's using `sort.Sort` rather than `sort.Stable` so it looks like you're right.